haystack_experimental/components/tools/openapi/_openapi.py

```diff
@@ -0,0 +1,557 @@
+        self.parameters = self.operation_dict.get(
+            "parameters", []
+        ) + self.spec_dict.get("paths", {}).get(self.path, {}).get("parameters", [])
+
+    def get_parameters(
+        self, location: Optional[Literal["header", "query", "path"]] = None
+    ) -> List[Dict[str, Any]]:
+        """
+        Get the parameters for the operation.
+
+        :param location: The location of the parameters to get.
+        :return: The parameters for the operation as a list of dictionaries.
+        """
+        if location:
+            return [param for param in self.parameters if param["in"] == location]
+        return self.parameters
+
+    def get_server(self, server_index: int = 0) -> str:
+        """
+        Get the servers for the operation.
+
+        :param server_index: The index of the server to use.
+        :return: The server URL.
+        :raises ValueError: If no servers are found in the specification.
+        """
+        servers = self.operation_dict.get("servers", []) or self.spec_dict.get(
+            "servers", []
+        )
+        if not servers:
+            raise ValueError("No servers found in the provided specification.")
+        if server_index >= len(servers):
+            raise ValueError(
+                f"Server index {server_index} is out of bounds. "
+                f"Only {len(servers)} servers found."
+            )
+        return servers[server_index].get(
+            "url"
+        )  # just use the first server from the list
+
+
+class OpenAPISpecification:
+    """
+    Represents an OpenAPI specification. See https://spec.openapis.org/oas/latest.html for details.
+    """
+
+    def __init__(self, spec_dict: Dict[str, Any]):
+        if not isinstance(spec_dict, Dict):
+            raise ValueError(
+                f"Invalid OpenAPI specification, expected a dictionary: {spec_dict}"
+            )
+        # just a crude sanity check, by no means a full validation
+        if (
+            "openapi" not in spec_dict
+            or "paths" not in spec_dict
+            or "servers" not in spec_dict
+        ):
+            raise ValueError(
+                "Invalid OpenAPI specification format. See https://swagger.io/specification/ for details.",
+                spec_dict,
+            )
+        self.spec_dict = spec_dict
+
+    @classmethod
+    def from_str(cls, content: str) -> "OpenAPISpecification":
+        """
+        Create an OpenAPISpecification instance from a string.
+
+        :param content: The string content of the OpenAPI specification.
+        :return: The OpenAPISpecification instance.
+        """
+        try:
+            loaded_spec = json.loads(content)
+        except json.JSONDecodeError:
+            try:
+                loaded_spec = yaml.safe_load(content)
+            except yaml.YAMLError as e:
+                raise ValueError(
+                    "Content cannot be decoded as JSON or YAML: " + str(e)
+                ) from e
```

Comment:
Let remove `from_str` for now.

Reply:
@shadeMe but this function is a must for parsing and is used from other `from_*` functions. I'll make it private

Reply:
Whoops, overlooked that entirely 🙈 Sorry, let's keep it as it is - no harm in keeping it a public function in that case. Sorry again!

Reply:
Resolved! Reverted to from_str